### PR TITLE
base: fioconfig: Support SOTA_CLIENT customization on aktualizr-toml-update handler.

### PR DESCRIPTION
**Pull request by caiotpereira**
This commit fixes the `fioconfig` handler allowing to restart of the  SOTA CLIENT service when the Custom SOTA Client is used instead `aktualizr-lite`.

Previously the `fioconfig` handler had a hard-coded `systemctl restart aktualizr-lite` . 
In cases where the Custom SOTA Client is used and the `fioconfig` receives any configuration from `fioctl` , it was impossible to restart the Custom SOTA Client service to apply the new configurations.

- Bump `fioconfig` rev to `c9469923f88cec496a9abce194d698fda0f3573d`.
- The final image still retains the `aktualizr-toml-update` handler filename.

### CODE REVIEW

**Review by Tim-Anderson — APPROVED** [review]
LGTM

**Review by MrCry0 — CHANGES_REQUESTED** [review]
@caiotpereira please squash the commit that fixes the brackets.

also, we use rebase strategy instead of merging. Could you rebase your branch on top of the upstream `main`? The resulting PR should contain only one commit.

**Comment by ricardosalveti** [comment]
Also please change the patch subject from meta-lmp-base to base only.

**Comment by vkhoroz** [comment]
@caiotpereira https://github.com/foundriesio/fioconfig/pull/41 was merged.

Please, update this PR accordingly, and I will approve it.

**Comment by mike-sul** [comment]
Also, this PR require another commit to bump the corresponding fioconfig version.

**Review by angolini — APPROVED** [review]
LGTM
when you are working on the other reviews, could you please add this sentence to the commit log?
"This pull request has been made based on the modifications found at https://github.com/foundriesio/fioconfig/commit/c9469923f88cec496a9abce194d698fda0f3573d."

**Review by quaresmajose — APPROVED** [review]

**Review by MrCry0 — APPROVED** [review]
LGTM.

In spite of a minor improvement that may be done:
I'd rewrite the commit message with an explanation of "why" instead of "what was done":
```
base: fioconfig: Support SOTA_CLIENT customization on aktualizr-toml-update handler

Bump fioconfig version and deploy a handler that supports customizing the SOTA_CLIENT.
```

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (vkhoroz) [inline meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:14]: Not sure if we usually do a _commit hash upgrade_ in a separate commit, or it is fine to combine it with code changes.

@ricardosalveti can give the guidance on this.

**Review by vkhoroz — APPROVED** [review]
LGTM

I like @MrCry0's idea. I'd make it even shorter:

```
base: fioconfig: Support SOTA_CLIENT customization on aktualizr-toml-update handler
```

The other sentence is _wordy details_.

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (vkhoroz) [inline meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:31]: No need for this black line, right?

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (ricardosalveti) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:14]: As in this case the only change involved is what allowed the recipe / packaging change, it is fine to have in a single commit.

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (ricardosalveti) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:31]: Yeah, please drop the extra line.

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (ricardosalveti) [inline meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Please perform this change on do_install:append as it it won't have any implication at the configure and compile steps (it is just a script file).

That way we can reuse the cache from configure and compile when changing SOTA_CLIENT.

Comment (ricardosalveti) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Also, there is no need to create a custom file, just perform sed -i instead.

**Comment by ricardosalveti** [comment]
Please also update the commit message:

`base: fioconfig: Bump SRC_URI and update the sed regex to replace the SOTA_CLIENT.`

Use `base: fioconfig: bump rev ...' instead.

SRC_URI is the same in the case, what got changed is the revision.

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (vkhoroz) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: @ricardosalveti we've discussed the `-i` thing in https://github.com/foundriesio/fioconfig/pull/41.

A problem with it is this: during development, once an `aktualizr-toml-update` file was updated in local environment, subsequent changes to SOTA_CLIENT have no effect (as the file content differs). The `-customized` intermediary mitigates this problem, as the original file remains unmodified.

Comment (vkhoroz) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: We can put the `-customized` thing into `/tmp`, but I have no idea what ramifications it might have; so IMHO it's not worth trying that.

Comment (ricardosalveti) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Oh sure, for local dev I wouldn't suggest using sed -i indeed, but with yocto it is usually better to avoid creating an extra file unless done as part of the respective build steps (e.g. generate a new file in do_compile and later install this new file with the common name under do_install). This is needed as yocto cache works per task, and usually do_config is done just to prepare do_compile, which can generate whatever content that is needed, and later do_install is responsible for moving stuff create as part of do_compile into the deploy folder (which is later transformed into packages based on FILES).

Comment (ricardosalveti) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Since here we're just replacing a script variable, we can do as do_install:append and reuse the task cache from configure and compile.

Comment (vkhoroz) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: I'll leave it to you guys, I'm personally fine either way. Just told you the concerns @caiotpereira expressed earlier regarding `-i`.

Comment (vkhoroz) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Doing this inside `do_install:append` definitely makes sense to me.

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (caiotpereira) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:14]: In this case, the SRCREV bump is aligned with the changes. since the changes are minor on `fioconfig` side and here, I don´t think it requires an extra PR.

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (caiotpereira) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Previously I was using a `template` folder to avoid this situation. 
@ricardosalveti @mike-sul  should we move the new handler to the `template` folder or manipulate the `FILES:${PN}` to include the correct files?
What do you think about your suggestion?

Comment (quaresmajose) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: > Hello @ricardosalveti, I noticed that you used sed -i for the test, but this could potentially cause problems for those who build locally and decide to change the SOTA_CLIENT after the first build. To avoid any cache issues, we have opted to keep the template file and copy it with the same name as the previous implementation.

if the `SOTA_CLIENT`  change after the build the task that uses this variable needs to run again and if this not happens we need to add this variable to the hash using the `do_task[vardeps] += "SOTA_CLIENT"`. I will test it locally to see if it works.

**Review by quaresmajose — APPROVED** [review]

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (quaresmajose) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Hello @ricardosalveti, I noticed that you used sed -i for the test, but this could potentially cause problems for those who build locally and decide to change the SOTA_CLIENT after the first build. To avoid any cache issues, we have opted to keep the template file and copy it with the same name as the previous implementation.

if the SOTA_CLIENT change after the build the task that uses this variable needs to run again and if this not happens we need to add this variable to the hash using the do_task[vardeps] += "SOTA_CLIENT". I will test it locally to see if it works.

Comment (quaresmajose) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: tested changing the install task and it works:
```
--- a/meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
+++ b/meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
@@ -11,7 +11,7 @@ SRC_URI = "git://${GO_IMPORT};protocol=${GO_IMPORT_PROTO};branch=main \
        file://fioconfig.path \
        file://fioconfig-extract.service \
 "
-SRCREV = "0b14975bbcfa85590432a9e71b1eb0125d81a92a"
+SRCREV = "c9469923f88cec496a9abce194d698fda0f3573d"
 
 UPSTREAM_CHECK_COMMITS = "1"
 
@@ -34,6 +34,8 @@ do_install:append() {
        install -m 0755 ${S}/src/${GO_IMPORT}/contrib/aktualizr-toml-update ${D}${datadir}/fioconfig/handlers
        install -m 0755 ${S}/src/${GO_IMPORT}/contrib/factory-config-vpn ${D}${datadir}/fioconfig/handlers
        install -m 0755 ${S}/src/${GO_IMPORT}/contrib/renew-client-cert ${D}${datadir}/fioconfig/handlers
+       sed -i -e "s/SOTA_CLIENT=aktualizr-lite/SOTA_CLIENT=${SOTA_CLIENT}/g" \
+               ${D}${datadir}/fioconfig/handlers/aktualizr-toml-update
 }
```

To check if the task will re-run whan changing SOTA_CLIENT we can:

1. Building the default with `bitbake fioconfig` to rebuild the package and the sstate-cache.
2. change SOTA_CLIENT to use a difent value: `echo 'SOTA_CLIENT = "aktualizr"' >> conf/auto.conf`
3. check the sigantures with: `bitbake fioconfig --dump-signatures=printdiff`

```
Task fioconfig:do_install couldn't be used from the cache because:
  We need hash 57ab39a5ea88d43f312e2510a162a32998d0cfc5106cf361c38be33220f6041a, closest matching task was 229ebc24493e77a08df14f3fc944059684b6431dbc2eaf8238fafbe3950f7a0b
  basehash changed from dbfe15beea7b6bc891c1ab9f3788e7a70ac1efd9e6187507a86b311086d3d5b6 to 30f9bb8bcde7eb0f8d539de1410bdb193f8b5607f2c5b1b7d67e36bd7bc3a63b
  Variable SOTA_CLIENT value changed from 'aktualizr-lite' to 'aktualizr'
```

**Review by quaresmajose — CHANGES_REQUESTED** [review]

File: meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb
Comment (vkhoroz) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: This looks like copy-on-write; clever and nice.

Comment (caiotpereira) [reply meta-lmp-base/recipes-support/fioconfig/fioconfig_git.bb:27]: Great @quaresmajose , thanks , nice idea!

**Review by quaresmajose — APPROVED** [review]
LGTM

**Comment by ricardosalveti** [comment]
Will merge after the release.

**Review by mike-sul — APPROVED** [review]

**Pull request merged**
